File: algorithm-binary-search-implementation-complete/src/main/java/com/binarysearch/algorithms/BruteForceSolutions.java
package com.binarysearch.algorithms;

/**
 * This class provides brute-force (typically O(N)) solutions
 * to problems that can be optimized with Binary Search.
 * These are included for comparison and to demonstrate the performance benefits
 * of Binary Search.
 */
public class BruteForceSolutions {

    /**
     * Brute-force solution for finding an element in an array.
     * Iterates through the array linearly.
     *
     * Time Complexity: O(N) - In the worst case, we traverse the entire array.
     * Space Complexity: O(1) - Uses a constant amount of extra space.
     *
     * @param nums   The array to search in.
     * @param target The element to find.
     * @return The index of the target if found, otherwise -1.
     */
    public int find(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == target) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Brute-force solution for finding the first occurrence of an element.
     * Iterates through the array linearly from the beginning.
     *
     * Time Complexity: O(N) - In the worst case, we traverse the entire array.
     * Space Complexity: O(1) - Uses a constant amount of extra space.
     *
     * @param nums   The array to search in.
     * @param target The element to find.
     * @return The index of the first occurrence of the target if found, otherwise -1.
     */
    public int findFirst(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == target) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Brute-force solution for finding the last occurrence of an element.
     * Iterates through the array linearly from the end.
     *
     * Time Complexity: O(N) - In the worst case, we traverse the entire array.
     * Space Complexity: O(1) - Uses a constant amount of extra space.
     *
     * @param nums   The array to search in.
     * @param target The element to find.
     * @return The index of the last occurrence of the target if found, otherwise -1.
     */
    public int findLast(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        for (int i = nums.length - 1; i >= 0; i--) {
            if (nums[i] == target) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Brute-force solution for searching in a rotated sorted array.
     * Iterates through the array linearly.
     *
     * Time Complexity: O(N) - In the worst case, we traverse the entire array.
     * Space Complexity: O(1) - Uses a constant amount of extra space.
     *
     * @param nums   The rotated sorted array.
     * @param target The element to find.
     * @return The index of the target if found, otherwise -1.
     */
    public int searchInRotatedSortedArray(int[] nums, int target) {
        return find(nums, target); // Linear scan is the brute force for this too
    }

    /**
     * Brute-force solution for finding a peak element.
     * Iterates through the array and checks each element against its neighbors.
     * Handles boundary conditions by assuming nums[-1] and nums[n] are -infinity.
     *
     * Time Complexity: O(N) - We might check every element.
     * Space Complexity: O(1) - Uses a constant amount of extra space.
     *
     * @param nums The array to search for a peak.
     * @return The index of a peak element. If multiple peaks exist, any one is acceptable.
     *         Returns -1 if array is empty.
     */
    public int findPeakElement(int[] nums) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        if (nums.length == 1) {
            return 0; // Single element is always a peak
        }

        for (int i = 0; i < nums.length; i++) {
            boolean isLeftLess = (i == 0) || (nums[i] > nums[i - 1]);
            boolean isRightLess = (i == nums.length - 1) || (nums[i] > nums[i + 1]);
            if (isLeftLess && isRightLess) {
                return i;
            }
        }
        return -1; // Should not happen based on problem constraints (peak always exists)
    }

    /**
     * Brute-force solution for Sqrt(x).
     * Iterates linearly from 0 up to x, checking squares.
     *
     * Time Complexity: O(sqrt(x)) - In the worst case, we iterate up to sqrt(x).
     * Space Complexity: O(1) - Uses a constant amount of extra space.
     *
     * @param x The non-negative integer.
     * @return The integer part of the square root of x.
     */
    public int mySqrt(int x) {
        if (x < 0) {
            throw new IllegalArgumentException("Input must be non-negative.");
        }
        if (x == 0) {
            return 0;
        }

        for (long i = 1; i <= x; i++) { // Use long for i*i to prevent overflow
            long square = i * i;
            if (square == x) {
                return (int) i;
            }
            if (square > x) {
                return (int) (i - 1);
            }
        }
        return x; // Fallback for very large x if loop condition allows (e.g., if x is Integer.MAX_VALUE and i exceeds int)
                  // For x=1, loop i=1, 1*1=1, return 1.
                  // For x=2, loop i=1, 1*1=1 < 2, continue. i=2, 2*2=4 > 2, return (2-1)=1.
    }
}

File: leetcode-binary-search-solutions-plus/src/algorithms/bruteForceSolutions.js
```javascript
/**
 * src/algorithms/bruteForceSolutions.js
 *
 * This file contains brute-force or naive implementations for some of the problems
 * covered in `binarySearchProblems.js`. These are primarily for comparison
 * during performance benchmarking to highlight the efficiency gains of binary search.
 */

// --- Problem 1: Standard Binary Search ---

/**
 * Standard Search - Brute Force (Linear Scan)
 * Finds the index of a target value in an array by iterating through all elements.
 *
 * @param {number[]} nums - The array of integers. (Sorting is not required for this approach).
 * @param {number} target - The integer to search for.
 * @returns {number} The index of the target if found, otherwise -1.
 *
 * Time Complexity: O(N)
 *   - In the worst case (target at the end or not present), it iterates through all N elements.
 * Space Complexity: O(1)
 */
function standardLinearSearch(nums, target) {
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] === target) {
      return i;
    }
  }
  return -1;
}

// --- Problem 2: Find First and Last Occurrence of an Element ---

/**
 * Finds the first and last occurrence of a target using brute force (linear scan).
 *
 * @param {number[]} nums - The sorted array of integers.
 * @param {number} target - The integer to search for.
 * @returns {number[]} `[first_occurrence_index, last_occurrence_index]` or `[-1, -1]`.
 *
 * Time Complexity: O(N)
 *   - Iterates through the entire array potentially twice (or once for finding both).
 * Space Complexity: O(1)
 */
function findFirstAndLastOccurrenceBruteForce(nums, target) {
  let firstIndex = -1;
  let lastIndex = -1;

  // Find first occurrence
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] === target) {
      firstIndex = i;
      break; // Found the first, exit loop
    }
  }

  // If target not found at all, return early
  if (firstIndex === -1) {
    return [-1, -1];
  }

  // Find last occurrence (can start from firstIndex for slight optimization, but worst case still O(N))
  for (let i = nums.length - 1; i >= 0; i--) {
    if (nums[i] === target) {
      lastIndex = i;
      break; // Found the last, exit loop
    }
  }

  return [firstIndex, lastIndex];
}

// --- Problem 3: Search in Rotated Sorted Array ---

/**
 * Searches for a target in a rotated sorted array using brute force (linear scan).
 *
 * @param {number[]} nums - The rotated sorted array of integers.
 * @param {number} target - The integer to search for.
 * @returns {number} The index of the target if found, otherwise -1.
 *
 * Time Complexity: O(N)
 * Space Complexity: O(1)
 */
function searchInRotatedSortedArrayBruteForce(nums, target) {
  return standardLinearSearch(nums, target); // Linear search works directly.
}

// --- Problem 4: Find Peak Element ---

/**
 * Finds a peak element using brute force (linear scan).
 *
 * @param {number[]} nums - The input array.
 * @returns {number} The index of a peak element.
 *
 * Time Complexity: O(N)
 *   - Iterates through the array to check each element against its neighbors.
 * Space Complexity: O(1)
 */
function findPeakElementBruteForce(nums) {
  if (nums.length === 0) return -1; // Or throw error, depending on problem spec
  if (nums.length === 1) return 0;

  // Check first element
  if (nums[0] > nums[1]) return 0;
  // Check last element
  if (nums[nums.length - 1] > nums[nums.length - 2]) return nums.length - 1;

  // Check elements in between
  for (let i = 1; i < nums.length - 1; i++) {
    if (nums[i] > nums[i - 1] && nums[i] > nums[i + 1]) {
      return i;
    }
  }
  return -1; // Should not happen given problem constraints (a peak always exists).
}

// --- Problem 5: Koko Eating Bananas ---

/**
 * Helper function to calculate total hours needed for a given speed `k`.
 * (Same as the optimized version's helper, as it's already optimal for a single `k`).
 */
function calculateHours(piles, k) {
  let hoursNeeded = 0;
  for (const pile of piles) {
    hoursNeeded += Math.ceil(pile / k);
  }
  return hoursNeeded;
}

/**
 * Finds the minimum eating speed `k` using brute force.
 * It iterates through all possible `k` values from 1 up to `max(piles)`
 * and checks if each `k` is feasible.
 *
 * @param {number[]} piles - An array of banana pile sizes.
 * @param {number} h - The maximum allowed hours.
 * @returns {number} The minimum integer `k`.
 *
 * Time Complexity: O(N * M)
 *   - Where N is the number of piles and M is the maximum pile size (range of `k`).
 *   - Iterates `M` times for `k`, and in each iteration, calls `calculateHours` which takes `O(N)`.
 * Space Complexity: O(1)
 */
function minEatingSpeedBruteForce(piles, h) {
  if (piles.length === 0) return 0;

  const maxPile = Math.max(...piles);
  // Iterate through all possible speeds from 1 up to the maximum pile size.
  for (let k = 1; k <= maxPile; k++) {
    if (calculateHours(piles, k) <= h) {
      return k; // The first `k` that satisfies the condition is the minimum.
    }
  }
  return maxPile; // Should not be reached if a solution is guaranteed.
}


module.exports = {
  standardLinearSearch,
  findFirstAndLastOccurrenceBruteForce,
  searchInRotatedSortedArrayBruteForce,
  findPeakElementBruteForce,
  minEatingSpeedBruteForce,
};
```